### ISSUE TITLE
Make the backend configurable

### ISSUE THREAD

**Issue by chanwit** [weaveworks/tf-controller#97]
Currently we support only Kubernetes backend.
But some users still require to use their existing backends like S3 or Dynamo tables outside Kubernetes.

It would be great to make the backend configurable to support those existing systems.

Here's a possible API design for custom backend.
```yaml
---
apiVersion: infra.contrib.fluxcd.io/v1alpha1
kind: Terraform
metadata:
  name: tf-controller-ng-cc5a1ac4
  namespace: flux-system
spec:
  interval: 1h
  path: ./_artifacts/20-zz-terraform
  approvePlan: "auto"
  backend:
    customConfiguration: |
      backend "s3" {
        bucket = "mybucket"
        key    = "path/to/my/key"
        region = "us-east-1"
      }
  varsFrom:
  - kind: Secret
    name: aws-vars
  sourceRef:
    kind: GitRepository
    name: flux-system
    namespace: flux-system
```

**Comment by fsequeira1** [comment]
Hi, I am interested in working on this

**Comment by chanwit** [comment]
Thank you @fsequeira1

**Comment by chanwit** [comment]
@fsequeira1 have you got a chance to tackle this issue please?

**Comment by fsequeira1** [comment]
@chanwit it is not forgotten but still haven't got an opportunity to work on it

**Comment by chanwit** [comment]
Thank you, @fsequeira1. 
I just wanted to check about it, as it's also in the roadmap of Q1, which is 15 days left. 

But I can postpone it to Q2 if you think it won't be landing by the Q1 milestone.
Please feel free to let me know which one works for you.

**Comment by fsequeira1** [comment]
@chanwit I think it's better to postpone to Q2 but I will try to deliver it asap

**Comment by chanwit** [comment]
Thank you @fsequeira1 
I'll move it to Q2.

**Comment by chanwit** [comment]
@fsequeira1 here's a design of the customDefinition block that could help you speeding things up.
Please define `customDefinition` as *string and if it's specified, you can dump the whole string to override the backend block, for example.

```
---
apiVersion: infra.contrib.fluxcd.io/v1alpha1
kind: Terraform
metadata:
  name: tf-controller-ng-cc5a1ac4
  namespace: flux-system
spec:
  backend:
    customConfiguration: |
      backend "s3" {
        bucket = "mybucket"
        key    = "path/to/my/key"
        region = "us-east-1"
      }
```

**Comment by o6uoq** [comment]
Assuming this implies Terraform Remote Backends, as listed [here](https://www.terraform.io/language/settings/backends) ?

If so, +1 for S3 and Consul as they are the most common backends in my experience with customers.

**Comment by chanwit** [comment]
Yes it is

**Comment by chanwit** [comment]
@fsequeira1 have you got any update on this please?

**Comment by fsequeira1** [comment]
@chanwit I can create a pr with the backend part, but it still lacks validation of the backend and support for secrets.

**Comment by chanwit** [comment]
fixed by #288
